File: tanzania/tanzania/report/wcf_contribution_report/wcf_contribution_report.js
// Copyright (c) 2024, Navari Limited and contributors
// For license information, please see license.txt

frappe.query_reports["WCF Contribution Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "employee",
			"label": __("Employee"),
			"fieldtype": "Link",
			"options": "Employee"
		},
		{
			"fieldname": "department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department"
		}
	],

	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Highlight WCF contribution column
		if (column.fieldname == "wcf_contribution" && data && data.wcf_contribution > 0) {
			value = "<span style='color:green; font-weight:bold'>" + value + "</span>";
		}

		// Highlight running total
		if (column.fieldname == "running_total" && data && data.running_total > 0) {
			value = "<span style='color:blue; font-weight:bold'>" + value + "</span>";
		}

		return value;
	},

	onload: function(report) {
		// Add custom button for WCF export
		report.page.add_inner_button(__("Export for WCF"), function() {
			frappe.msgprint(__("Exporting report for WCF submission..."));
			// TODO: Add custom export logic for WCF submission
		});

		// Add button to show summary
		report.page.add_inner_button(__("Show Summary"), function() {
			let data = report.data;
			if (data && data.length > 0) {
				let total_wcf = 0;

				data.forEach(row => {
					total_wcf += row.wcf_contribution || 0;
				});

				frappe.msgprint({
					title: __("WCF Contribution Summary"),
					message: `
						<table class="table table-bordered">
							<tr>
								<td><strong>Total Employees:</strong></td>
								<td>${data.length}</td>
							</tr>
							<tr>
								<td><strong>Total WCF Contribution (0.5%):</strong></td>
								<td><strong>${format_currency(total_wcf)}</strong></td>
							</tr>
							<tr>
								<td colspan="2">
									<em>Note: WCF is employer-only contribution (0.5% of gross pay)</em>
								</td>
							</tr>
						</table>
					`,
					indicator: 'green'
				});
			}
		});
	}
};

File: tanzania/tanzania/report/wcf_contribution_report/wcf_contribution_report.py
# Copyright (c) 2026, Nelson Mpanju and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data


def get_columns():
	"""Return columns for WCF Contribution Report"""
	return [
		{
			"fieldname": "employee",
			"label": _("Employee ID"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": 120
		},
		{
			"fieldname": "employee_name",
			"label": _("Employee Name"),
			"fieldtype": "Data",
			"width": 200
		},
		{
			"fieldname": "wcf_number",
			"label": _("WCF Number"),
			"fieldtype": "Data",
			"width": 130
		},
		{
			"fieldname": "designation",
			"label": _("Designation"),
			"fieldtype": "Link",
			"options": "Designation",
			"width": 150
		},
		{
			"fieldname": "department",
			"label": _("Department"),
			"fieldtype": "Link",
			"options": "Department",
			"width": 150
		},
		{
			"fieldname": "date_of_joining",
			"label": _("Date of Joining"),
			"fieldtype": "Date",
			"width": 120
		},
		{
			"fieldname": "gross_pay",
			"label": _("Gross Pay"),
			"fieldtype": "Currency",
			"width": 130
		},
		{
			"fieldname": "wcf_contribution",
			"label": _("WCF Contribution (0.5%)"),
			"fieldtype": "Currency",
			"width": 160
		},
		{
			"fieldname": "running_total",
			"label": _("Running Total"),
			"fieldtype": "Currency",
			"width": 150
		}
	]


def get_data(filters):
	"""Get WCF contribution data for all employees (employer-only contribution)"""
	if not filters:
		filters = {}

	conditions = get_conditions(filters)

	# Get salary slip data for all employees
	salary_slips = frappe.db.sql("""
		SELECT
			ss.name,
			ss.employee,
			ss.employee_name,
			ss.department,
			ss.designation,
			ss.gross_pay,
			ss.posting_date,
			e.date_of_joining,
			e.wcf_number
		FROM `tabSalary Slip` ss
		INNER JOIN `tabEmployee` e ON ss.employee = e.name
		WHERE ss.docstatus = 1
	""" + conditions + """
		ORDER BY ss.employee, ss.posting_date
	""", filters, as_dict=1)

	data = []
	running_total = 0

	for slip in salary_slips:
		# Calculate WCF contribution (0.5% of gross pay - employer only)
		wcf_contribution = flt(slip.gross_pay) * 0.005

		# Update running total
		running_total += wcf_contribution

		row = {
			"employee": slip.employee,
			"employee_name": slip.employee_name,
			"wcf_number": slip.wcf_number or "",
			"designation": slip.designation,
			"department": slip.department,
			"date_of_joining": slip.date_of_joining,
			"gross_pay": flt(slip.gross_pay),
			"wcf_contribution": flt(wcf_contribution),
			"running_total": flt(running_total)
		}

		data.append(row)

	return data


def get_conditions(filters):
	"""Build SQL conditions from filters"""
	conditions = ""

	if filters.get("company"):
		conditions += " AND ss.company = %(company)s"

	if filters.get("from_date"):
		conditions += " AND ss.start_date >= %(from_date)s"

	if filters.get("to_date"):
		conditions += " AND ss.end_date <= %(to_date)s"

	if filters.get("employee"):
		conditions += " AND ss.employee = %(employee)s"

	if filters.get("department"):
		conditions += " AND ss.department = %(department)s"

	return conditions
